### PR TITLE
 refactor(iam): cleanup of IAM library

### PR DESCRIPTION
**Pull request by rix0rrr**
Various changes.

* Move PolicyStatement to its own file.
* Remove `postProcess` from Token interface, move opting into it into `IResolvable.resolve()`, so it can be hidden from the `PolicyDocument` public interface.
* Introduce `PolicyStatement.fromAttributes({ ... })` as an alternative to the fluent interface chaining of old `PolicyStatement`.

Plus all the breaking changes below:

BREAKING CHANGE:

* **iam**: `PolicyStatement` no longer has a fluid API, and accepts a
  props object to be able to set the important fields.
* **iam**: rename `ImportedResourcePrincipal` to `UnknownPrincipal`.
* **iam**: `managedPolicyArns` renamed to `managedPolicies`, takes
  return value from `ManagedPolicy.fromAwsManagedPolicyName()`.
* **iam**: `PolicyDocument.postProcess()` is now removed.
* **iam**: `PolicyDocument.addStatement()` renamed to `addStatements`.
* **iam**: `PolicyStatement` is no longer `IResolvable`
* **iam**:  `AwsPrincipal` has been removed, use `ArnPrincipal` instead.

----

### Pull Request Checklist

* [ ] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [ ] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
  - __Design__: For significant features, design document added to `design` folder
* [ ] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** and module name in parens, which will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [ ] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/aws-iam/lib/policy-statement.ts [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+  /**
+   * Render the policy statement to JSON
+   */
+  toStatementJson(): any;
+}
+
+/**
+ * Represents a statement in an IAM policy document.
+ */
+export class PolicyStatement implements IPolicyStatement {
```
Comment (eladb) [inline packages/@aws-cdk/aws-iam/lib/policy-statement.ts]: Should we make a linter rule to enforce the usage of IPolicyStatement everywhere?

File: packages/@aws-cdk/aws-iam/lib/policy.ts [hunk 2]
```diff
@@ -118,15 +119,15 @@ export class Policy extends Resource implements IPolicy {
...
-      props.statements.forEach(p => this.addStatement(p));
+      props.statements.forEach(p => this.addStatements(p));
     }
   }
 
   /**
    * Adds a statement to the policy document.
    */
-  public addStatement(statement: PolicyStatement) {
-    this.document.addStatement(statement);
```
Comment (eladb) [inline packages/@aws-cdk/aws-iam/lib/policy.ts]: IPolicyStatement?

File: packages/@aws-cdk/cdk/lib/private/resolve.ts [hunk 3]
```diff
@@ -36,12 +36,18 @@ export function resolve(obj: any, options: IResolveOptions): any {
   /**
    * Make a new resolution context
    */
-  function makeContext(appendPath?: string): IResolveContext {
+  function makeContext(appendPath?: string): [IResolveContext, IPostProcessor] {
```
Comment (RomainMuller) [inline packages/@aws-cdk/cdk/lib/private/resolve.ts]: Can we avoid returning a tuple, and instead return some interface? The proliferation of `[0]` pains me...

Comment (rix0rrr) [reply packages/@aws-cdk/cdk/lib/private/resolve.ts]: It's private though.

Comment (RomainMuller) [reply packages/@aws-cdk/cdk/lib/private/resolve.ts]: I still have to read all those `[0]` :)

**Pull request merged**